BUG: Fix three `complex`- & `float128`-related issues with `nd_grid`

BvB93 opened the pull request:
This pull request fixes three issues with the `np.nd_grid.__getitem__()` method, 
more specifically a set of discrepancies encountered whenever a `slice` object is passed versus a tuple of `slice` objects. 
These two syntaxes will henceforth be referred to as, respectively, the "tuple of slices"- and "slice"-syntax.

```python
import numpy as np

np.mgrid[0:10]  # "slice" syntax
np.mgrid[0:10,]  # "tuple of slices" syntax
```

To summarize:
* First observed in https://github.com/numpy/numpy/issues/16945 different data types are encountered between the two syntaxes when a `np.float128` instance is passed to `start`, `stop` and/or `step`. This issue has been fixed; the "tuple of slices" syntax now uses the same dtype conversion rules as `np.arange()` (which was already used for the "slice" syntax).
``` python
>>> np.mgrid[0:10:np.float128(0.1)]
array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float128)

>>> np.mgrid[0:10:np.float128(0.1),]
array([[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]])

```
* The "tuple of slices" syntax failed to return a `float128` array if `stop` is a `float128` instance and `step` is a complex number.
``` python
>>> np.mgrid[0:np.float128(10):10j]
array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
```
* Non-identical arrays are returned if `step` is a complex number whose absolute value is non-integer. 
  The documentation ([docs](https://numpy.org/doc/stable/reference/generated/numpy.mgrid.html)) specifies that `start` and `stop` should be included, 
  _i.e._ the "tuple-of-slices" syntax produces the correct result.
``` python

>>> np.mgrid[0:10:2.5j]
array([0.        , 6.66666667])

>>> np.mgrid[0:10:2.5j,]
array([[ 0., 10.]])
```

Commit:
TST: Update the `np.mgrid` tests

Commit:
BUG: Fixed an issue where `nd_grid` could cast `float128` to `float64`

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -262,6 +269,25 @@ def test_accepts_npfloating(self):
         assert_(grid32.dtype == np.float64)
         assert_array_almost_equal(grid64, grid32)
 
+    @pytest.mark.skipif(FLOAT128_EX, reason=str(FLOAT128_EX))
```
If you use `np.longdouble` instead of `np.float128`, you don't need this marker.

BvB93 replied:
Good to know; implemented in https://github.com/numpy/numpy/pull/16971/commits/22789628d1cfe5e2ddb29a4466cbcd2224c85b67.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -262,6 +269,25 @@ def test_accepts_npfloating(self):
+        grid128c_a = mgrid[0:np.float128(1):3.4j]
+        grid128c_b = mgrid[0:np.float128(1):3.4j,]
+        assert_(grid128c_a.dtype == grid128c_b.dtype == np.float128)
+        assert_array_almost_equal(grid128c_a, grid128c_b[0])
```
I'd expect these to be exactly equal, not just almost equal

BvB93 replied:
I feel it would be wise to stick to `assert_array_almost_equal()` considering we're dealing with float arrays.
just in case to deal with any numerical noise that might creep in.

eric-wieser replied:
"noise" does not creep into deterministic calculations like this - computing a 1d vs 2d grid of numbers should result in identical bitpatterns, else we have another bug.

BvB93 replied:
Ah yes, I missed the fact that this comparison involved two arrays of the same dtype.
Addressed in https://github.com/numpy/numpy/pull/16971/commits/2e94e7e9f54efb17be9a59f14f55deae1e6e758b.

BvB93 commented:
On another note, the `nd_grid` class currently has quite a few opportunities for code style improvements.
If there are no complaints then I'd like to give it a shot in a follow up pull request.

eric-wieser commented on numpy/lib/index_tricks.py:
```diff
@@ -195,12 +196,12 @@ def __getitem__(self, key):
-                step = abs(step)
-                length = int(step)
+                # Prevent the (potential) creation of integer arrays
+                step = length = _nx.floor(abs(step))
```
Why `floor` instead of `int`?

BvB93 replied:
If `step == 1` then this could lead to a situation where an integer array is returned (if `start` and `stop` are integers), while previously it would _always_ return a float array.
`floor()` ensures that `step` remains a float.
I suppose `float(int(abs(step)))` would also do the trick.

mattip replied:
@BvB93 do you intend to change this?

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/16971/commits/4dcc8408efdffc6181bc67cc72fe6741b3eeae1d.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -262,6 +262,24 @@ def test_accepts_npfloating(self):
         assert_(grid32.dtype == np.float64)
         assert_array_almost_equal(grid64, grid32)
 
+    def test_accepts_npfloating128(self):
```
```suggestion
    def test_accepts_longdouble(self):
```

eric-wieser commented on numpy/lib/index_tricks.py:
```diff
@@ -146,24 +146,25 @@ def __init__(self, sparse=False):
             size = []
-            typ = int
+            # Mimick the behavior of `np.arange` and use a data type
+            # which is at least as large as `int` / `np.long`
```
```suggestion
            # which is at least as large as `np.int_`
```
`np.long` is deprecated

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/16971/commits/e55436361436bf3cd024e54610d068974b89cc89.

eric-wieser commented on numpy/lib/index_tricks.py:
```diff
@@ -195,12 +196,12 @@ def __getitem__(self, key):
                     step = (key.stop-start)/float(step-1)
-                stop = key.stop + step
-                return _nx.arange(0, length, 1, float)*step + start
+                typ = _nx.result_type(0.0, start, stop, step)
```
```suggestion
                typ = _nx.result_type(0, start, stop, step)
```
for consistency with the no-comma case. Almost certainly makes no difference, so all the more reason to be consistent.

eric-wieser replied:
Wait nevermind, I think the bug is the other way around.

BvB93 replied:
This does actually make a difference here. 
As to remain consisting with numpy `<= 1.19` the return dtype _must_ be float when `step` is a complex number. 
The complex `step` is converted into an integer a few lines earlier, so without the `0.0` this could lead to the unwanted situation where `result_type()` returns an integer type.

BvB93 replied:
> Wait nevermind, I think the bug is the other way around.

The other way around is still fine btw, as `step` is never converted into an integer when determining the return dtype.

``` python
                if isinstance(step, (_nx.complexfloating, complex)):
                    step = abs(step)
                    size.append(int(step))
```

eric-wieser commented on numpy/lib/index_tricks.py:
```diff
@@ -146,24 +146,25 @@ def __init__(self, sparse=False):
                     step = 1
                 if isinstance(step, (_nx.complexfloating, complex)):
-                    size.append(int(abs(step)))
-                    typ = float
```
With this line removed, what is `np.mgrid[1:2:1j,].dtype`, and does it match `np.mgrid[1:2:1j].dtype?

BvB93 replied:
It does, but it doesn't seem to be explicitly tested anywhere. 
I'll add one in a moment.

Examples
---------
``` python
In [1]: import numpy as np

In [2]: np.mgrid[1:2:1j,].dtype
Out[2]: dtype('float64')

In [3]: np.mgrid[1:2:1j].dtype
Out[3]: dtype('float64')
```

BvB93 replied:
Oh hang on, I did actually add a test for test:

``` python

        # Related to #16945
        assert_array_equal(
            mgrid[0.1:0.3:3.3j], mgrid[0.1:0.3:3.3j,][0],
        )
```

eric-wieser commented on numpy/lib/index_tricks.py:
```diff
@@ -146,24 +146,25 @@ def __init__(self, sparse=False):
         try:
             size = []
-            typ = int
+            # Mimick the behavior of `np.arange` and use a data type
```
```suggestion
            # Mimic the behavior of `np.arange` and use a data type
```

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/16971/commits/8a5e1d58b8e57819a9d33aa356dbef43f690cfd3.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -262,6 +262,24 @@ def test_accepts_npfloating(self):
+        grid128c_b = mgrid[0:np.longdouble(1):3.4j,]
+        assert_(grid128c_a.dtype == grid128c_b.dtype == np.longdouble)
+        assert_array_equal(grid128c_a, grid128c_b[0])
+
```
Lets add one more of these:
```suggestion

        grid128c_a = mgrid[0:1:np.clongdouble(3.4j]
        grid128c_b = mgrid[0:1:np.clongdouble(3.4j),]
        assert_(grid128c_a.dtype == grid128c_b.dtype == np.longdouble)
        assert_array_equal(grid128c_a, grid128c_b[0])
```

BvB93 replied:
This was a good call actually. `grid128c_a` had the `float64` dtype while `grid128c_b` was of `float128` 
(which one would expect as `type(abs(np.complex256(...))) is np.float128`).

Anyway, it's been fixed as of https://github.com/numpy/numpy/pull/16971/commits/c093dcb90d5c5de1e7d8af0ea2681567316cf0e0 and the tests have been updated in https://github.com/numpy/numpy/pull/16971/commits/3a49c7c750fddc2093700d8d844155be86c0f302.

eric-wieser replied:
I think this was the bug I was trying to catch above.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -273,6 +291,12 @@ def test_accepts_npcomplexfloating(self):
         )
 
+        # Related to #16945
+        assert_array_equal(
```
As far as I remember, `assert_array_equal` does not check the return type, only the value

BvB93 replied:
Yup, you're right.
I added a few explicit dtype checks in https://github.com/numpy/numpy/pull/16971/commits/3a49c7c750fddc2093700d8d844155be86c0f302.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -273,6 +291,19 @@ def test_accepts_npcomplexfloating(self):
+        assert_(grid64_a.dtype == grid64_b.dtype == np.float64)
+        assert_array_equal(grid64_a, grid64_b)
+
+        float_type = np.result_type(int, np.longdouble)
+        grid128_a = mgrid[0.1:0.3:np.clongdouble(3.3j)]
+        grid128_b = mgrid[0.1:0.3:np.clongdouble(3.3j),][0]
+        assert_(grid128_a.dtype == grid128_b.dtype == float_type)
```
Any reason for this and not
```suggestion
        float_type = np.result_type(int, np.longdouble)
        grid128_a = mgrid[0.1:0.3:np.clongdouble(3.3j)]
        grid128_b = mgrid[0.1:0.3:np.clongdouble(3.3j),][0]
        assert_(grid128_a.dtype == grid128_b.dtype == np.longdouble)
```

BvB93 replied:
This is specifically for windows: 
`np.result_type(int, np.longdouble) is np.float64` as opposed to `... is np.longdouble`.

eric-wieser replied:
Not for me:
```
In [173]: np.result_type(int, np.longdouble) is np.float64
Out[173]: False

In [174]: np.__version__
Out[174]: '1.17.4'
```
Did this change recently?

BvB93 replied:
Ah, correction: it returns `dtype(float64)`, not just `float64`:

``` python
In [1]: import numpy as np

In [2]: np.result_type(int, np.longdouble)
Out[2]: dtype('float64')
```

eric-wieser replied:
```
In [1]: np.result_type(int, np.longdouble)
Out[1]: dtype('float64')

In [2]: np.result_type(int, np.longdouble).type is np.float64
Out[2]: False
```
Can you confirm that on master that `False` is `True`?

BvB93 replied:
> Can you confirm that on master that the `False` and `True` swap?

False alarm. the string notation of `np.dtype(np.longlong)` was the cause of confusion here.
The `np.result_type(int, np.longdouble).type is np.longdouble` expression indeed holds for the master, even for windows.

Reverted in https://github.com/numpy/numpy/pull/16971/commits/e548b9a7812607c12ab6afa5a7f659484d29af69.

BvB93 commented:
Rebased and squashed all commits:
* https://github.com/numpy/numpy/pull/16971/commits/42c5512e2dbcccf1821ad8995841eb67dff4336f BUG: Fixed an issue where `nd_grid` could cast `float128` to `float64`
* https://github.com/numpy/numpy/pull/16971/commits/34ba7d7db0a43060f430e84d072f114faad80993 TST: Update the `np.mgrid` tests

charris commented:
@BvB93 Is this still needed?

BvB93 commented:
> @BvB93 Is this still needed?

@charris the issues outlined in this PR still seem to be present, so I'd say this PR would still be a worthwhile addition.

charris commented:
Thanks Bas.
